ASSERTION FAILED: !m_objectStoresByIdentifier.contains(objectStore->info().identifier()) in MemoryIDBBackingStore.cpp

szewai opened the pull request:
#### 0a51ad5e5ed55fd363616a9f0bd788767db3b523
<pre>
ASSERTION FAILED: !m_objectStoresByIdentifier.contains(objectStore-&gt;info().identifier()) in MemoryIDBBackingStore.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=250908">https://bugs.webkit.org/show_bug.cgi?id=250908</a>
rdar://104163599

Reviewed by Chris Dumez.

m_deletedObjectStores in MemoryBackingStoreTransaction is used to track deleted object stores that need to be added
back to database if transaction is aborted, so it should not contain object stores created in the transaction. If an
object store is created and deleted in the same transaction, database should not have information about the object store
after transaction finishes.

* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort-expected.txt: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort-private-expected.txt: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort-private.html: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort.html: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-rename-abort-expected.txt:
* LayoutTests/storage/indexeddb/modern/objectstore-rename-abort-private-expected.txt:
* LayoutTests/storage/indexeddb/modern/resources/objectstore-add-delete-abort.js: Copied from LayoutTests/storage/indexeddb/modern/resources/objectstore-rename-abort.js.
(openDatabase):
(onDatabaseUpgradeNeeded):
(onTransactionAbort):
* LayoutTests/storage/indexeddb/modern/resources/objectstore-rename-abort.js:
(onDatabaseUpgradeNeeded):
* Source/WebCore/Modules/indexeddb/server/MemoryBackingStoreTransaction.cpp:
(WebCore::IDBServer::MemoryBackingStoreTransaction::objectStoreDeleted):

Canonical link: <a href="https://commits.webkit.org/259215@main">https://commits.webkit.org/259215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/535bfc7ff99e75b4bed854aea10c06ebbd9112ab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13226 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113343 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173641 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4141 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96363 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112408 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109905 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38678 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92860 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25682 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80336 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6592 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27055 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6729 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3600 "Found 3 new test failures: fast/images/avif-as-image.html, fast/images/avif-heif-container-as-image.html, storage/indexeddb/modern/deleteindex-4-private.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46594 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6354 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8510 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
Please make sure these new tests are passing in Chrome (and ideally in Firefox) to make sure our behavior is interoperable.

szewai committed:
ASSERTION FAILED: !m_objectStoresByIdentifier.contains(objectStore->info().identifier()) in MemoryIDBBackingStore.cpp
https://bugs.webkit.org/show_bug.cgi?id=250908
rdar://104163599

Reviewed by Chris Dumez.

m_deletedObjectStores in MemoryBackingStoreTransaction is used to track deleted object stores that need to be added
back to database if transaction is aborted, so it should not contain object stores created in the transaction. If an
object store is created and deleted in the same transaction, database should not have information about the object store
after transaction finishes.

* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort-expected.txt: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort-private-expected.txt: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort-private.html: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-add-delete-abort.html: Added.
* LayoutTests/storage/indexeddb/modern/objectstore-rename-abort-expected.txt:
* LayoutTests/storage/indexeddb/modern/objectstore-rename-abort-private-expected.txt:
* LayoutTests/storage/indexeddb/modern/resources/objectstore-add-delete-abort.js: Copied from LayoutTests/storage/indexeddb/modern/resources/objectstore-rename-abort.js.
(openDatabase):
(onDatabaseUpgradeNeeded):
(onTransactionAbort):
* LayoutTests/storage/indexeddb/modern/resources/objectstore-rename-abort.js:
(onDatabaseUpgradeNeeded):
* Source/WebCore/Modules/indexeddb/server/MemoryBackingStoreTransaction.cpp:
(WebCore::IDBServer::MemoryBackingStoreTransaction::objectStoreDeleted):

Canonical link: https://commits.webkit.org/259215@main
